nifi-nar-bundles/nifi-deeplearning4j-bundle/nifi-deeplearning4j-processors/src/main/java/org/apache/nifi/processors/deeplearning4j/DeepLearning4JPredictor.java

```diff
@@ -0,0 +1,218 @@
+            String fieldSeparator = context.getProperty(FIELD_SEPARATOR).evaluateAttributeExpressions(flowFile).getValue();
+            String recordSeparator = context.getProperty(RECORD_SEPARATOR).evaluateAttributeExpressions(flowFile).getValue();
+
+            int [] dimensions = getInputDimensions(context, charset, flowFile, fieldSeparator);
+
+            if ( getLogger().isDebugEnabled() )    {
+                getLogger().debug("Received input {} with dimensions {}", new Object[] { input, dimensions });
+            }
+
+            MultiLayerNetwork model = getModel(context);
+
+            long startTimeMillis = System.currentTimeMillis();
+
+            String [] inputRecords = input.split(recordSeparator);
+
+            List<INDArray> features = Arrays.stream(inputRecords).map(
+                record -> {
+                    double [] parameters = Arrays.stream(record.split(fieldSeparator)).mapToDouble(
+                             field -> Double.parseDouble(field)).toArray();
+
+                    INDArray featureInput = Nd4j.create(parameters, dimensions);
+
+                    if ( getLogger().isDebugEnabled() ) {
+                        getLogger().debug("Features for record {} parameters {} dims {} featureInput {} ",
+                            new Object[] {record, parameters, dimensions, featureInput});
+                    }
+
+                    return featureInput;
+
+                }).collect(Collectors.toList());
+
+           INDArray allFeatures = Nd4j.vstack(features);
+
+           INDArray results = model.output(allFeatures);
+
+           double [][] partitionedResults = new double[inputRecords.length][];
+           for (int row = 0; row < inputRecords.length; row++) {
+                INDArray result = results.getRow(row);
+                partitionedResults[row] = Nd4j.toFlattened(result).toDoubleVector();
+           }
+
+           String jsonResult = gson.toJson(partitionedResults);
+           int [] shape = results.shape();
+           String jsonShape = gson.toJson(Arrays.copyOfRange(shape, 1, shape.length));
+
+           if ( getLogger().isDebugEnabled() ) {
+                getLogger().debug("Prediction for inputRecords {}, dims {}, results {}, result.shape {}, partitionedResults {}, jsonResult {}, shape {}, jsonShape {}",
+                        new Object[] {inputRecords, dimensions, results, Arrays.toString(results.shape()), partitionedResults, jsonResult, shape, jsonShape});
+           }
+
+           try (ByteArrayInputStream bais = new ByteArrayInputStream(jsonResult.getBytes(charset))) {
+               flowFile = session.importFrom(bais, flowFile);
+           }
+
+           session.putAttribute(flowFile, DEEPLEARNING4J_OUTPUT_SHAPE, jsonShape);
+
+           final long endTimeMillis = System.currentTimeMillis();
+
+           session.transfer(flowFile, REL_SUCCESS);
+
+           session.getProvenanceReporter().send(flowFile, makeProvenanceUrl(context),
```

Comment:
I don't believe a SEND event is appropriate here. We are not sending the data to some external location. I think we want a CONTENT_MODIFIED event. It probably does make sense, though, to include the time duration and if wanting to include the name of the Model File, that could be done in the event details.

Reply:
Corrected.